code/modules/modular_computers/file_system/programs/airestorer.dm

```diff
@@ -1,127 +1,139 @@
-		return
-
-	if(cardhold.flush)
-		ai_slot.locked = FALSE
+	var/mob/living/silicon/ai/A = stored_card.AI
+	if(stored_card.flush)
 		restoring = FALSE
 		return
-	ai_slot.locked = TRUE
 	A.adjustOxyLoss(-5, FALSE)
 	A.adjustFireLoss(-5, FALSE)
 	A.adjustBruteLoss(-5, FALSE)
 
 	// Please don't forget to update health, otherwise the below if statements will probably always fail.
 	A.updatehealth()
-
 	if(A.health >= 0 && A.stat == DEAD)
 		A.revive(full_heal = FALSE, admin_revive = FALSE)
-		cardhold.update_appearance()
-
+		stored_card.update_appearance()
 	// Finished restoring
 	if(A.health >= 100)
-		ai_slot.locked = FALSE
 		restoring = FALSE
 
 	return TRUE
 
+/datum/computer_file/program/ai_restorer/try_insert(obj/item/attacking_item, mob/living/user)
+	if(!computer)
+		return FALSE
+	if(!istype(attacking_item, /obj/item/aicard))
+		return FALSE
 
-/datum/computer_file/program/aidiag/ui_data(mob/user)
-	var/list/data = get_header_data()
-	var/mob/living/silicon/ai/AI = get_ai()
+	if(stored_card)
+		to_chat(user, span_warning("You try to insert \the [attacking_item] into \the [computer.name], but the slot is occupied."))
+		return FALSE
+	if(user && !user.transferItemToLoc(attacking_item, computer))
+		return FALSE
+
+	stored_card = attacking_item
+	to_chat(user, span_notice("You insert \the [attacking_item] into \the [computer.name]."))
+
+	return TRUE
+
+/datum/computer_file/program/ai_restorer/try_eject(mob/living/user, forced = FALSE)
+	if(!stored_card)
+		if(user)
+			to_chat(user, span_warning("There is no card in \the [computer.name]."))
+		return FALSE
+
+	if(restoring && !forced)
+		if(user)
+			to_chat(user, span_warning("Safeties prevent you from removing the card until reconstruction is complete..."))
+		return FALSE
+
+	if(user && computer.Adjacent(user))
+		to_chat(user, span_notice("You remove [stored_card] from [computer.name]."))
+		user.put_in_hands(stored_card)
+	else
+		stored_card.forceMove(computer.drop_location())
+
+	stored_card = null
+	restoring = FALSE
+	return TRUE
 
-	var/obj/item/aicard/aicard = get_ai(2)
+
+/datum/computer_file/program/ai_restorer/ui_act(action, params)
+	. = ..()
+	if(.)
+		return
+
+	switch(action)
+		if("PRG_beginReconstruction")
+			if(!stored_card || !stored_card.AI)
+				return FALSE
+			var/mob/living/silicon/ai/A = stored_card.AI
+			if(A && A.health < 100)
+				restoring = TRUE
+				A.notify_ghost_cloning("Your core files are being restored!", source = computer)
+			return TRUE
+		if("PRG_eject")
+			if(stored_card)
+				try_eject(usr)
+				return TRUE
+
+/datum/computer_file/program/ai_restorer/ui_data(mob/user)
+	var/list/data = get_header_data()
 
 	data["ejectable"] = TRUE
-	data["AI_present"] = FALSE
+	data["AI_present"] = !!stored_card?.AI
 	data["error"] = null
-	if(!aicard)
+
+	if(!stored_card)
 		data["error"] = "Please insert an intelliCard."
+	else if(!stored_card.AI)
+		data["error"] = "No AI located"
+	else if(stored_card.flush)
+		data["error"] = "Flush in progress"
```

Comment:
```suggestion
	else if(!stored_card.AI)
		data["error"] = "No AI located!"
	else if(stored_card.flush)
		data["error"] = "Flush in progress..."
```

Reply:
this is the opposite bro..
no ai isn't something to exclaim about, it's just a generic error
AI being deleted as we speak however, is something to exclaim.

Reply:
i just thought some punctuation was better than nothing